File: csharp/say/Say.cs
public static class Say
{
    private static readonly List<string> small = [
        "zero",
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
        "ten",
        "eleven",
        "twelve",
        "thirteen",
        "fourteen",
        "fifteen",
        "sixteen",
        "seventeen",
        "eighteen",
        "nineteen",
    ];
    private static readonly List<string> tens = ["", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"];
    private static readonly List<string> scales = ["", "thousand", "million", "billion"];

    public static string InEnglish(long number)
    {
        if (number < 0 || number > 999_999_999_999) throw new ArgumentOutOfRangeException();

        if (number == 0) return "zero";

        var num = number;
        var scaleIndex = 0;
        List<string> parts = [];

        while (num > 0)
        {
            var chunk = (int)(num % 1000);

            if (chunk != 0)
            {
                var chunkText = SayUnder1000(chunk);
                var scale = scales[scaleIndex];

                parts.Add(scale == "" ? chunkText : $"{chunkText} {scale}");
            }

            num /= 1000;
            scaleIndex++;
        }

        parts.Reverse();

        return string.Join(" ", parts);
    }

    private static string SayUnder1000(int num)
    {
        List<string> parts = [];

        var hundreds = num / 100;
        var remainder = num % 100;

        if (hundreds > 0)
        {
            parts.Add($"{small[hundreds]} hundred");
        }

        if (remainder > 0)
        {
            if (remainder < 20)
            {
                parts.Add(small[remainder]);
            }
            else if (remainder % 10 == 0)
            {
                parts.Add(tens[remainder / 10]);
            }
            else
            {
                parts.Add($"{tens[remainder / 10]}-{small[remainder % 10]}");
            }
        }

        return string.Join(" ", parts);
    }
}

File: cpp/say/say.cpp
#include "say.h"
#include <array>
#include <vector>
#include <stdexcept>

namespace say
{

    static const std::array<std::string, 20> small{
        "zero",
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
        "ten",
        "eleven",
        "twelve",
        "thirteen",
        "fourteen",
        "fifteen",
        "sixteen",
        "seventeen",
        "eighteen",
        "nineteen"};

    static const std::array<std::string, 10> tens{
        "",
        "",
        "twenty",
        "thirty",
        "forty",
        "fifty",
        "sixty",
        "seventy",
        "eighty",
        "ninety"};

    static const std::array<std::string, 4> scales{
        "",
        "thousand",
        "million",
        "billion"};

    std::string say_under_1000(long num)
    {
        std::vector<std::string> parts;

        long hundreds = num / 100;
        long remainder = num % 100;

        if (hundreds > 0)
        {
            parts.push_back(std::string(small[hundreds]) + " hundred");
        }

        if (remainder > 0)
        {
            if (remainder < 20)
            {
                parts.push_back(small[remainder]);
            }
            else if (remainder % 10 == 0)
            {
                parts.push_back(tens[remainder / 10]);
            }
            else
            {
                parts.push_back(std::string(tens[remainder / 10]) + "-" + small[remainder % 10]);
            }
        }

        std::string result;

        for (size_t i = 0; i < parts.size(); i++)
        {
            if (i > 0)
            {
                result += ' ';
            }

            result += parts[i];
        }

        return result;
    }

    std::string in_english(long number)
    {
        if (number < 0 || number > 999999999999L)
            throw std::domain_error("Out of range");

        if (number == 0)
            return "zero";

        long num = number;
        int scale_index = 0;

        std::vector<std::string> parts;

        while (num > 0)
        {
            int chunk = (int)(num % 1000);

            if (chunk != 0)
            {
                std::string chunk_text = say_under_1000(chunk);
                std::string scale = scales[scale_index];

                parts.push_back(scale == "" ? chunk_text : chunk_text + " " + scale);
            }

            num /= 1000;
            scale_index++;
        }

        std::string result;

        for (auto it = parts.rbegin(); it != parts.rend(); ++it)
        {
            if (!result.empty())
            {
                result += ' ';
            }

            result += *it;
        }

        return result;
    }

} // namespace say
